ntp-client/src/Network/NTP/Client.hs

```diff
@@ -1,318 +1,85 @@
+    -> TVar NtpStatus
+    -> IO ()
+ntpClientThread tracer ntpSettings ntpStatus = forM_ restartDelay $ \t -> do
+    traceWith tracer $ NtpTraceRestartDelay t
+    awaitPendingWithTimeout ntpStatus $ t * 1_000_000
+    traceWith tracer NtpTraceRestartingClient
+    catchIOError queryLoop (\err -> traceWith tracer $ NtpTraceIOError err)
+    atomically $ writeTVar ntpStatus NtpSyncUnavailable
     where
-    closeSockets :: Sockets -> IO ()
-    closeSockets sockets = do
-        case sockets of
-            This a -> fn a
-            That a -> fn a
-            These a b -> fn a >> fn b
-        traceWith tracer NtpTraceSpawnNtpClientSocketsClosed
-
-    fn :: Last (WithAddrFamily t Socket.Socket) -> IO ()
-    fn (Last sock) = Socket.close $ runWithAddrFamily sock
-
-    -- Try to resolve addresses, on failure wait 30s and start again.
-    resolve = do
-        traceWith tracer NtpTraceSpawnNtpClientResolveDNS
-        addrs <- catMaybes <$> traverse (resolveNtpHost tracer) (ntpServers settings)
-        if null addrs
-          then do
-            atomically $ writeTVar ncStatus NtpSyncUnavailable
-            -- either wait 30s or wait for `NtpSyncPending` which might be set
-            -- by a client (e.g. wallet), in which case try to resolve the dns.
-            race_
-              (threadDelay 30000000)
-              (do
-                atomically $ do
-                  s <- readTVar ncStatus
-                  case s of
-                    NtpSyncPending -> return ()
-                    _              -> retry
-                traceWith tracer NtpTraceSpawnNtpClientResolvePending
-              )
-            resolve
-          else return addrs
-
--- Try to create IPv4 and IPv6 socket.
-mkSockets :: Tracer IO NtpTrace -> NtpClientSettings -> IO Sockets
-mkSockets tracer settings =
-    doMkSockets `catch` handleIOException >>= \case
-        Just socks -> pure socks
-        Nothing     -> do
-            traceWith tracer NtpTraceMkSocketsNoSockets
---            logWarning "Couldn't create either IPv4 or IPv6 socket, retrying in 5 sec..."
-            threadDelay 5000000
-            mkSockets tracer settings
-  where
-    doMkSockets :: IO (Maybe Sockets)
-    doMkSockets = do
-        addrs <- udpLocalAddresses
-        (<>) <$> (createAndBindSock tracer IPv4 addrs)
-             <*> (createAndBindSock tracer IPv6 addrs)
-
-    handleIOException :: IOException -> IO (Maybe Sockets)
-    handleIOException e = do
-        traceWith tracer $ NtpTraceMkSocketsIOExecption e
-        threadDelay 5000000
-        doMkSockets
+        restartDelay :: [Int]
+        restartDelay = [0, 5, 10, 20, 60, 180, 600] ++ repeat 600
+
+        queryLoop = ntpQuery tracer ntpSettings >>= \case
+            status@(NtpDrift _ ) -> do
+                atomically $ writeTVar ntpStatus status
+                traceWith tracer NtpTraceClientSleeping
+                awaitPendingWithTimeout ntpStatus $ fromIntegral $ ntpPollDelay ntpSettings
```

Comment:
this `queryLoop` will send requests much more often than `ntpPollDelay`, which is not our intention.
 
At this point of execution, `ntpStatus` will be likely (e.g. in good network conditions) set to `NtpDrift`, since `ntpQuery` awaits on results, and thus `awaitPendingWithTimeout` will return immediately.  Even if `ntpQuery` did not await on results, this would await only as long as the responses are in-flight, not as long as the `ntpPollDelay` instructs.

Reply:
I think it is correct as is it.
The intention is that `awaitPendingWithTimeout` waits until `ntpStatus` gets set to `NtpSyncPending` or the timeout.
`Network.NTP.Test.testClient` also works as expected.

Reply:
You're right, it's ok.